File: src/test/java/model/DeckUtilsTest.java
package model;

import com.aau.wizard.model.*;
import com.aau.wizard.model.enums.CardSuit;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class DeckUtilsTest {

    @Test
    void getTrumpSuit_returnsWizardSuit() {
        List<ICard> cards = List.of(
                CardFactory.createCard(CardSuit.RED, 5),
                CardFactory.createCard(CardSuit.BLUE, 14),
                CardFactory.createCard(CardSuit.GREEN, 7)
        );

        CardSuit trumpSuit = DeckUtils.getTrumpSuit(cards);
        assertEquals(CardSuit.BLUE, trumpSuit);
    }

    @Test
    void getTrumpSuit_returnsFirstWizardSuit() {
        List<ICard> cards = List.of(
                CardFactory.createCard(CardSuit.RED, 14),
                CardFactory.createCard(CardSuit.BLUE, 14),
                CardFactory.createCard(CardSuit.GREEN, 7)
        );

        CardSuit trumpSuit = DeckUtils.getTrumpSuit(cards);
        assertEquals(CardSuit.RED, trumpSuit);
    }

    @Test
    void getTrumpSuit_returnsNullWhenNoWizard() {
        List<ICard> cards = List.of(
                CardFactory.createCard(CardSuit.RED, 5),
                CardFactory.createCard(CardSuit.BLUE, 7),
                CardFactory.createCard(CardSuit.SPECIAL, 0)
        );

        CardSuit trumpSuit = DeckUtils.getTrumpSuit(cards);
        assertNull(trumpSuit);
    }

    @Test
    void getTrumpSuit_handlesEmptyList() {
        List<ICard> emptyList = List.of();
        CardSuit trumpSuit = DeckUtils.getTrumpSuit(emptyList);
        assertNull(trumpSuit);
    }

    @Test
    void isWizard_returnsTrueForWizard() {
        ICard wizard = CardFactory.createCard(CardSuit.SPECIAL, 14);
        assertTrue(WizardCard.isWizard(wizard));
    }

    @Test
    void isWizard_returnsFalseForNonWizard() {
        ICard numberCard = CardFactory.createCard(CardSuit.RED, 5);
        ICard jester = CardFactory.createCard(CardSuit.SPECIAL, 0);

        assertFalse(WizardCard.isWizard(numberCard));
        assertFalse(WizardCard.isWizard(jester));
    }

    @Test
    void isJester_returnsTrueForJester() {
        ICard jester = CardFactory.createCard(CardSuit.SPECIAL, 0);
        assertTrue(JesterCard.isJester(jester));
    }

    @Test
    void isJester_returnsFalseForNonJester() {
        ICard numberCard = CardFactory.createCard(CardSuit.BLUE, 7);
        ICard wizard = CardFactory.createCard(CardSuit.SPECIAL, 14);

        assertFalse(JesterCard.isJester(numberCard));
        assertFalse(JesterCard.isJester(wizard));
    }

    @Test
    void constructor_isPrivate() {
        assertThrows(IllegalAccessException.class, () -> {
            DeckUtils.class.getDeclaredConstructor().newInstance();
        });
    }
}

File: src/test/kotlin/com/aau/wizard/core/cards/DeckUtilsTest.kt
package com.aau.wizard.core.cards

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.api.Nested
import org.junit.jupiter.params.provider.EnumSource

class DeckUtilsTest {

    @Nested
    inner class GetTrumpSuitTests {
        @Test
        fun `returns null for empty list`() {
            assertNull(DeckUtils.getTrumpSuit(emptyList()))
        }

        @Test
        fun `returns null when no wizards present`() {
            val cards = listOf(
                Card(Suit.RED, 1),
                Card(Suit.YELLOW, 0), // Jester
                Card(Suit.BLUE, 5)
            )
            assertNull(DeckUtils.getTrumpSuit(cards))
        }

        @ParameterizedTest
        @EnumSource(Suit::class)
        fun `returns suit of first wizard found`(suit: Suit) {
            val cards = listOf(
                Card(Suit.GREEN, 5),
                Card(suit, 14), // Wizard
                Card(Suit.BLUE, 14) // Another wizard (should be ignored)
            )
            assertEquals(suit, DeckUtils.getTrumpSuit(cards))
        }

        @Test
        fun `works with multiple wizards of different suits`() {
            val expectedSuit = Suit.YELLOW
            val cards = listOf(
                Card(Suit.RED, 1),
                Card(expectedSuit, 14), // First wizard
                Card(Suit.BLUE, 14)    // Second wizard
            )
            assertEquals(expectedSuit, DeckUtils.getTrumpSuit(cards))
        }
    }

    @Nested
    inner class IsWizardTests {
        @Test
        fun `returns true for wizard card`() {
            assertTrue(DeckUtils.isWizard(Card(Suit.SPECIAL, 14)))
        }

        @Test
        fun `returns false for non-wizard cards`() {
            assertFalse(DeckUtils.isWizard(Card(Suit.RED, 1)))
            assertFalse(DeckUtils.isWizard(Card(Suit.YELLOW, 0))) // Jester
        }

        @ParameterizedTest
        @EnumSource(Suit::class)
        fun `returns true for wizard in any suit`(suit: Suit) {
            assertTrue(DeckUtils.isWizard(Card(suit, 14)))
        }
    }

    @Nested
    inner class IsJesterTests {
        @Test
        fun `returns true for jester card`() {
            assertTrue(DeckUtils.isJester(Card(Suit.SPECIAL, 0)))
        }

        @Test
        fun `returns false for non-jester cards`() {
            assertFalse(DeckUtils.isJester(Card(Suit.BLUE, 5)))
            assertFalse(DeckUtils.isJester(Card(Suit.GREEN, 14))) // Wizard
        }

        @ParameterizedTest
        @EnumSource(Suit::class)
        fun `returns true for jester in any suit`(suit: Suit) {
            assertTrue(DeckUtils.isJester(Card(suit, 0)))
        }
    }
}